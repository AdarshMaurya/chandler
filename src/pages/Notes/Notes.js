import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {View, FlatList, ActivityIndicator, Text} from 'react-native';
import moment from 'moment';

import {styles} from './Notes.styles';
import {Navbar, EmptyActivity} from 'components';
import {commonStyles} from 'theme';

export class Notes extends PureComponent {
  static propTypes = {
    back: PropTypes.func.isRequired,
    getNotesByContact: PropTypes.func.isRequired,
    notes: PropTypes.array.isRequired,
    isFetching: PropTypes.bool.isRequired,
  };

  componentWillMount() {
    this.props.getNotesByContact();
  }

  keyExtractor = (item, index) => String(item.id);

  renderFooter = () => {
    const {isFetching} = this.props;

    if (!isFetching) {
      return null;
    }

    return (
      <ActivityIndicator style={commonStyles.activityIndicator} size="large" />
    );
  };

  renderItem = ({item, index}) => {
    const {notes} = this.props;
    const note = notes[index];

    return (
      <View style={styles.noteContainer}>
        <Text style={styles.date}>{moment(note.created_at).format('LL')}</Text>
        <Text>{note.body}</Text>
      </View>
    );
  };

  render() {
    const {back, notes, getNotesByContact, isFetching} = this.props;

    return (
      <View style={commonStyles.flex}>
        <Navbar title="Notes" onBack={back} />
        {isFetching || notes.length ? (
          <FlatList
            data={notes}
            renderItem={this.renderItem}
            keyExtractor={this.keyExtractor}
            ListFooterComponent={this.renderFooter}
            onEndReached={getNotesByContact}
            onEndReachedThreshold={0.5}
          />
        ) : (
          <EmptyActivity
            image={require('./assets/empty-notes.png')}
            title="Notes are great to record precious details about your contacts."
            subtitle="Just write down what you have in mind."
          />
        )}
      </View>
    );
  }
}
