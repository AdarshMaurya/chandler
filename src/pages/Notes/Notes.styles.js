import {StyleSheet, PixelRatio} from 'react-native';

export const styles = StyleSheet.create({
  noteContainer: {
    paddingVertical: 15,
    paddingHorizontal: 30,
    backgroundColor: 'white',
    borderColor: '#E2E4E5',
    borderBottomWidth: 1 / PixelRatio.get(),
  },
  date: {
    fontSize: 12,
    marginBottom: 8,
  },
});
