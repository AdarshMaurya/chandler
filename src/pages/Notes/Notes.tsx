import moment from 'moment';
import React, {PureComponent} from 'react';
import {ActivityIndicator, FlatList, Text, View} from 'react-native';

import {EmptyActivity, Navbar} from '@components';
import {INote} from '@models';
import {IRouterBackOperation} from '@redux/router';
import {commonStyles} from '@theme';
import {styles} from './Notes.styles';

interface INotesProps {
  back: IRouterBackOperation;
  getNotesByContact: () => void;
  notes: INote[];
  isFetching: boolean;
}

export class Notes extends PureComponent<INotesProps, {}> {
  public componentWillMount() {
    this.props.getNotesByContact();
  }

  public keyExtractor = (item, index) => String(item.id);

  public renderFooter = () => {
    const {isFetching} = this.props;

    if (!isFetching) {
      return null;
    }

    return (
      <ActivityIndicator style={commonStyles.activityIndicator} size="large" />
    );
  };

  public renderItem = ({item, index}) => {
    const {notes} = this.props;
    const note = notes[index];

    return (
      <View style={styles.noteContainer}>
        <Text style={styles.date}>{moment(note.created_at).format('LL')}</Text>
        <Text>{note.body}</Text>
      </View>
    );
  };

  public render() {
    const {back, notes, getNotesByContact, isFetching} = this.props;

    return (
      <View style={commonStyles.flex}>
        <Navbar title="Notes" onBack={back} />
        {isFetching || notes.length ? (
          <FlatList
            data={notes}
            renderItem={this.renderItem}
            keyExtractor={this.keyExtractor}
            ListFooterComponent={this.renderFooter}
            onEndReached={getNotesByContact}
            onEndReachedThreshold={0.5}
          />
        ) : (
          <EmptyActivity
            image={require('./assets/empty-notes.png')}
            title="Notes are great to record precious details about your contacts."
            subtitle="Just write down what you have in mind."
          />
        )}
      </View>
    );
  }
}
