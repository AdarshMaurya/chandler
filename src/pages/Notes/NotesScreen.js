import {connect} from 'react-redux';

import {Notes} from './Notes';
import {back} from 'redux/router';
import {getNotesByContact} from 'redux/notes';

export const mapStateToProps = (state, {navigation}) => {
  const contact = state.contacts[navigation.state.params];
  return {
    notes: (contact.notes || []).map(noteId => state.notes[noteId]),
    isFetching: state.getNotesByContact.isFetching,
  };
};

export const mapDispatchToProps = (dispatch, {navigation}) => ({
  back: () => dispatch(back()),
  getNotesByContact: () => dispatch(getNotesByContact(navigation.state.params)),
});

export const NotesScreen = connect(mapStateToProps, mapDispatchToProps)(Notes);
