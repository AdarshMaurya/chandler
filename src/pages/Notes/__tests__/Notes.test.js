import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import {Notes} from '../Notes';

describe('Pages', () => {
  describe('Notes', () => {
    let back;
    let getNotesByContact;

    beforeEach(() => {
      back = jest.fn();
      getNotesByContact = jest.fn();
    });

    it('should renders correctly without notes and fetching', () => {
      const notes = [];
      expect(
        toJson(
          shallow(
            <Notes
              back={back}
              getNotesByContact={getNotesByContact}
              isFetching={true}
              notes={notes}
            />,
          ),
        ),
      ).toMatchSnapshot();
    });

    it('should renders correctly without notes and not fetching', () => {
      const notes = [];
      expect(
        toJson(
          shallow(
            <Notes
              back={back}
              getNotesByContact={getNotesByContact}
              isFetching={false}
              notes={notes}
            />,
          ),
        ),
      ).toMatchSnapshot();
    });

    it('should renders correctly with notes and fetching', () => {
      const notes = [{body: 'My body', created_at: '1991-12-17'}];
      expect(
        toJson(
          shallow(
            <Notes
              back={back}
              getNotesByContact={getNotesByContact}
              isFetching={true}
              notes={notes}
            />,
          ),
        ),
      ).toMatchSnapshot();
    });

    it('should renders correctly with notes and fetching', () => {
      const notes = [
        {body: 'My body', created_at: '1991-12-17'},
        {body: 'My body 2', created_at: '1993-12-17'},
      ];
      expect(
        toJson(
          shallow(
            <Notes
              back={back}
              getNotesByContact={getNotesByContact}
              isFetching={false}
              notes={notes}
            />,
          ),
        ),
      ).toMatchSnapshot();
    });
  });
});
