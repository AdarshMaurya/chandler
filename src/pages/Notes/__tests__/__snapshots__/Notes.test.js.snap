// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages Notes should renders correctly with notes and fetching 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Notes"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={
      Array [
        Object {
          "body": "My body",
          "created_at": "1991-12-17",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`Pages Notes should renders correctly with notes and fetching 2`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Notes"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={
      Array [
        Object {
          "body": "My body",
          "created_at": "1991-12-17",
        },
        Object {
          "body": "My body 2",
          "created_at": "1993-12-17",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`Pages Notes should renders correctly without notes and fetching 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Notes"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`Pages Notes should renders correctly without notes and not fetching 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Notes"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;
